Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP1Connections.swift

```diff
@@ -0,0 +1,602 @@
+        }
+
+        enum FailedConnectionAction {
+            /// Creates a new connection, that needs to be started
+            case replace
+            /// Removes the old connection.
+            case remove
+        }
+
+        /// Remove or replace a connection.
+        ///
+        /// Call this function after the connection start backoff is done, or a live connection is closed.
+        ///
+        /// - Parameters:
+        ///   - connectionID: The failed connection's ID
+        ///   - closure: A closure in which a decision is made how to progress with this connection spot
+        ///
+        /// - Returns: A new connectionID and eventLoop to create a new connection, if the action is `.replace`
+        mutating func failConnection(
+            _ connectionID: Connection.ID,
+            _ closure: (FailedConnectionContext) -> (FailedConnectionAction)
+        ) -> (Connection.ID, EventLoop)? {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to create a new connection that we know nothing about?")
+            }
+
+            let use: ConnectionUse
+            let eventLoop = self.connections[index].eventLoop
+            let starting: Int
+            if index < self.overflowIndex {
+                use = .generalPurpose
+                // we need to subtract 1, since we dont want this connection to count
+                starting = self.startingGeneralPurposeConnections - 1
+            } else {
+                use = .eventLoop(eventLoop)
+                // we need to subtract 1, since we dont want this connection to count
+                starting = self.startingEventLoopConnections(on: eventLoop) - 1
+            }
+
+            let context = FailedConnectionContext(
+                eventLoop: eventLoop,
+                use: use,
+                connectionsStartingForUseCase: starting
+            )
+
+            switch closure(context) {
+            case .remove:
+                if index < self.overflowIndex {
+                    self.overflowIndex = self.connections.index(before: self.overflowIndex)
+                }
+                self.connections.remove(at: index)
+                return nil
+            case .replace:
+                let newConnection = HTTP1ConnectionState(
+                    connectionID: self.generator.next(),
+                    eventLoop: self.connections[index].eventLoop
+                )
+
+                self.connections[index] = newConnection
+                return (newConnection.connectionID, newConnection.eventLoop)
+            }
+        }
+
+        // MARK: Shutdown
+
+        mutating func shutdown() -> CleanupContext {
+            var cleanupContext = CleanupContext()
+            let initialOverflowIndex = self.overflowIndex
+
+            self.connections = self.connections.enumerated().compactMap { index, connectionState in
```

Comment:
What's special about the connections that we keep track of?

Reply:
documented in the cleanup call of a single state

Reply:
Much clearer with the cleanup action too!